### DISCUSSION TITLE
How do I improve the correction with varying illumination?

### DISCUSSION THREAD

**Discussion by danoneata** [colour-science/colour-checker-detection#67] [Q&A] [answered]
Dear Thomas,

In the [example tutorial](https://github.com/colour-science/colour-checker-detection/blob/master/colour_checker_detection/examples/examples_detection.ipynb), you showcase the color calibration algorithm for a couple of images. I was wondering, did you use a special illuminant (_e.g._, D65) when taking these pictures? If not, how come that the calibration works so well given that the color calibration methods are sensitive to the illumination conditions, as you very eloquently argue in [one of your posts](https://www.colour-science.org/posts/the-colorchecker-considered-mostly-harmless/)? My question is prompted by some experiments I did, where I took a picture of the color checker outdoors during midday, but in somewhat cloudly conditions
<img src="https://user-images.githubusercontent.com/819256/192530841-409859d3-b1c8-49c0-b718-041d3cea579c.png" width=500/>
And I obtain results that are visually worse than yours:
- before
![image](https://user-images.githubusercontent.com/819256/192531045-d75493af-16f8-4632-af44-1aefe7c7fa3e.png)
- after color correction (linear model)
![image](https://user-images.githubusercontent.com/819256/192531105-9fee2b6d-c8f9-4c97-bb1a-f93209e19fbd.png)

Do you think this is due to the illumination?

Thank you,
– Dan

**Comment by KelSolaar** [comment, marked as answer]
Hi @danoneata,

So I spent a bit of time looking at your image, something that is throwing off the computations is that you have an alpha channel so please make sure that you are taking only the RGB channels: `colour.read_image(path)[..., 0:3]`. The other issue is that it seems like that the image has some kind of contrast curve applied. How was it processed?

Cheers,

Thomas

**Reply by danoneata** [reply]
Ah, I think that Github converted the image to PNG when I've uploaded it and added the alpha channel. The [original image](https://doneata.bitbucket.io/static/PXL_20220927_112002854.jpg) has three components:
```
PXL_20220927_112002854.jpg: JPEG image data, JFIF standard 1.01, aspect ratio, density 72x72, segment length 16, Exif Standard: [TIFF image data, big-endian, direntries=10, manufacturer=Google, model=Pixel 4a, orientation=upper-left, xresolution=152, yresolution=160, resolutionunit=2, software=HDR+ 1.0.471358035zd, datetime=2022:09:27 14:20:02, GPS-Data], baseline, precision 8, 4032x3024, components 3
```

But maybe the problem is that I've used the JPG image and not the raw one? Maybe the camera software applied by default a contrast curve? I'll redo the experiment tomorrow using the raw capture.

Thanks a lot for sharing your thoughts!

**Reply by danoneata** [reply]
Hello! I've redone the experiment by using [a raw image](https://doneata.bitbucket.io/static/PXL_20220930_084436202.dng) which I convert to `PPM` using a command similar to the one in the tutorial:
```
dcraw -g 2.4 12.92 PXL_20220930_084436202.dng
```
For comparison I'm also including the results [on the JPG-captured image](https://doneata.bitbucket.io/static/PXL_20220930_084436202.jpg) and I display them below—JPG (left) and RAW → PPM (right) images; second row shows the colors before correction and the third row after. While the neutrals are better now compared to the JPG version, the other color are still worse than what is show in the tutorial example. Do you think this is the best achievable result or am I still doing something wrong?

For reference the code for color correction is [here](https://gist.github.com/danoneata/7e9cf4133e6753ecef6c40ad2c980ac4), but it is very similar to the one in the tutorial.

Thank you!

![image](https://user-images.githubusercontent.com/819256/193243227-e7639dd1-ee9c-40e0-b524-ace02ddc2cd8.png)

**Reply by KelSolaar** [reply]
There seem to be a non-insignificant amount of specular response on the chart and probably vignette too. You might be reaching the extent of what a 3x3 colour correction matrix can do and need to rely on higher-order functions. They do come with teeth though, anything outside the training data, i.e. the colour checker, might suffer a lot. If you can explain what you are trying to achieve in the end, I might be able to give you some advices.

**Reply by danoneata** [reply]
Thanks a lot, Thomas!

Yes, I'm also bit uneasy of applying non-linear transformations since they can get rather unstable from my experiments.

This is example was just a sanity check to make sure that what I'm doing is reasonable. The practical case is closer to what I've described in [the other thread](https://github.com/colour-science/colour/discussions/747#discussioncomment-3733906). More precisely, we are capturing pictures of wood planks in a controlled environment (with artificial light and using [Allied Vision's Manta](https://www.alliedvision.com/en/support/technical-documentation/) cameras), but it seems that the color of the wood in the photos are somewhat different from what we see with bare eye in reality. So we thought that applying a color correction transform will alleviate this behavior (but it looks like that things are more subtle than that and that we can easily end up doing more harm than good if we don't know what we are doing). Of course, any advice is greatly appreciated!

**Reply by KelSolaar** [reply]
If you are in a controlled environment, I would strongly recommend using a spectroradiometer to measure the lighting and use that to produce reference values for the colour rendition chart, you really have the perfect conditions to do that.

**Reply by KelSolaar** [reply]
I'm also working with @aforsythe and we have made a tool to calibrate blackbox/prosumer cameras available here: https://github.com/ampas/idt-calculator
